notebooks/value_semantics/01_value_semantics_in_swift.ipynb

```diff
@@ -0,0 +1,949 @@
+    "    }\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "extension Bar: CustomStringConvertible {\n",
+    "    public var description: String {\n",
+    "        \"Bar(f: \\(f), y: \\(y))\"\n",
+    "    }\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "b: Bar(f: Foo(x: 0), y: 0)\r\n",
+      "b: Bar(f: Foo(x: 1), y: 1) (post-baz())\r\n",
+      "c: Bar(f: Foo(x: 1), y: 1) (create copy)\r\n",
+      "b: Bar(f: Foo(x: 1), y: 1) (should be equivalent to above)\r\n",
+      "b: Bar(f: Foo(x: 2), y: 1) (f should be incremented)\r\n",
+      "c: Bar(f: Foo(x: 1), y: 1) (should be the same as original c value)\r\n",
+      "b: Bar(f: Foo(x: 3), y: 2) (more changes)\r\n",
+      "c: Bar(f: Foo(x: 1), y: 1) (more sames)\r\n"
+     ]
+    }
+   ],
+   "source": [
+    "func testFunc3() {\n",
+    "    var b = Bar()\n",
+    "    print(\"b: \\(b)\")\n",
+    "    b.baz()\n",
+    "    print(\"b: \\(b) (post-baz())\")\n",
+    "    let c = b\n",
+    "    print(\"c: \\(c) (create copy)\")\n",
+    "    print(\"b: \\(b) (should be equivalent to above)\")\n",
+    "    b.f.incrementByOne()\n",
+    "    print(\"b: \\(b) (f should be incremented)\")\n",
+    "    print(\"c: \\(c) (should be the same as original c value)\")\n",
+    "    b.baz()\n",
+    "    print(\"b: \\(b) (more changes)\")\n",
+    "    print(\"c: \\(c) (more sames)\")\n",
+    "//     c.baz() // Compile error\n",
+    "//     c.f.incrementByOne()  // Also compile error\n",
+    "}\n",
+    "testFunc3()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Swift's non-obvious properties\n",
+    "\n",
+    "I've collected a few nice features of Swift that aren't apparent by reading [the language guide](https://docs.swift.org/swift-book/LanguageGuide/TheBasics.html).\n",
+    "\n",
+    "Said another way, these properties of Swift enable high performance execution without sacrificing value semantics.\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Early dropping of references\n",
+    "\n",
+    "Swift does not have a defined point of destruction for local variables, and thus the compiler is free to destroy variables early. Swift doesn't specify the memory layout of a local stack frame, which allows for a variety of optimizations, such as hoisting values to registers or tearing them apart. Below are a few code examples of this."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "/// An Int, but implemented using the same lazy copy-on-write techniques as\n",
+    "/// arrays (and Tensors) for demonstration purposes (makes copies explicit).\n",
+    "public struct CoWInt {\n",
```

Comment:
IMO this example isn't illustrating anything about exploding values/local stack frames as claimed above, and keeping the copy count in the instance makes for a complex example that's hard for users to analyze.  We should discuss how to improve this part.

One idea: maybe it would be better simply to have `CoWInt` print a message when it makes a copy

Reply:
I think that would be a good improvement.